### PR TITLE
BEAM-3243 better error message when there are conflicting anonymous names

### PR DESCRIPTION
**Pull request by rmannibucau**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 1]
```diff
@@ -578,12 +586,51 @@ void validate(PipelineOptions options) {
...
-        case ERROR:
+        case ERROR: // be very verbose here since it will just fail the execution
           throw new IllegalStateException(
               String.format(
                   "Pipeline update will not be possible"
                       + " because the following transforms do not have stable unique names: %s.",
-                  Joiner.on(", ").join(unstableNames)));
+                  Joiner.on(", ").join(unstableNames)) + "\n\n"
+                      + "Conflicting instances:\n"
+                      + Joiner.on("\n").join(Iterables.transform(unstableNames,
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Hmm I'm having a hard time following this code for two reasons:
- I don't remember in what format are the unstable names here. It would help to add some comments with examples of intermediate values at each step
- Java7 syntax makes the functional programming style here very verbose, I think this would be more readable if refactored into simple loops and static utility functions (ideally with their own unit tests)

File: sdks/java/core/src/test/java/org/apache/beam/sdk/PipelineTest.java [hunk 2]
```diff
@@ -117,6 +121,50 @@ public PipelineResult run(Pipeline pipeline) {
...
+    thrown.expectMessage(new BaseMatcher<String>() { // more readable than a regex
+      @Override
+      public void describeTo(final Description description) {
+        description.appendText("validates the conflicting instances are "
+                + "listed into the exception message");
+      }
+
+      @Override
+      public boolean matches(final Object o) {
+        final String sanitized = String.class.cast(o)
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/PipelineTest.java]: Can you add a comment with an example of an actual value here?

**Comment by rmannibucau** [comment]
Updated the PR, hope it addresses the comments.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 3]
```diff
@@ -636,4 +649,49 @@ public void visitPrimitiveTransform(Node node) {
       node.getTransform().validate(options);
     }
   }
+
+  private static class TransformToMessage implements Function<PTransform<?, ?>, String> {
+    @Override
+    public String apply(final PTransform<?, ?> transform) {
+      final Object representant;
+      if (ParDo.SingleOutput.class.isInstance(transform)) {
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Hmm why are we special-casing ParDo here - why is line 664 ok for other transforms but not OK for ParDo.Single/MultiOutput? If it's a matter of the transform's .toString() - can you just change the toString() methods on these ParDo classes if they're currently not good enough?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 4]
```diff
@@ -636,4 +649,49 @@ public void visitPrimitiveTransform(Node node) {
...
+
+    @Override
+    public String apply(final String input) {
+      Collection<PTransform<?, ?>> values = null;
+      String currentName = input;
+      // strip the counter appended to the name,
+      // see uniquifyInternal().
+      // common example: ParDo(Anonymous) becoming ParDo(Anonymous)2
+      while (values == null && !currentName.isEmpty()) {
+        currentName = currentName.substring(0, currentName.length() - 1);
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: What if the counter is greater than 10, and takes more than 1 digit?

In general this seems brittle: this is just extracting the original non-unique name of a transform, right? Can't we pass the offending PTransform's here directly somehow, e.g. at line 532 key things by the original (non-unique) name and traverse entries that have more than 1 transform? I think then you can also just remove the variable "unstableNames" because it's redundant w.r.t. the map.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 3 above)
Comment (rmannibucau) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: To get a user data and not an internal beam data. ParDo doesnt help here for instance.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 4 above)
Comment (rmannibucau) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: The loop will remove both digits normally.

Thought about removing it but then you need to browse the whole collection instead of just testing if the list is empty - twice. No strong opinion on it but thought that holding the direct information "conflict" was valuable.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 3 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Sorry, I don't understand which internal Beam data you have in mind. Line 664 calls toString() - so all you need to get rid of special-casing ParDo as opposed to the dozens of other Beam transforms is to add .toString() methods to ParDo.SingleOutput and MultiOutput, which will be useful for debugging anyway.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 4 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: I think traversing the collection is much better than parsing digit-by-digit a string that we ourselves have formatted.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 3 above)
Comment (rmannibucau) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Hmm, not really cause the goal is to log the min info needed for the user to fix it. XOutput doesnt help since it is ignored in the name computation.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 4 above)
Comment (rmannibucau) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Will change it tomorrow

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 3 above)
Comment (jkff) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: If I understand correctly: the default .toString() on a ParDo returns simply "ParDo.SingleOutput@something" or alike; and instead you want to return a .toString() of the DoFn inside it. Can you just override the toString() method on the ParDo.SingleOutput class to say "return fn.toString()"?

Comment (rmannibucau) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Sounds a bit weird to bypass a wrapper but Im fine with it. Will push it tmr. Thanks for the comment/details.

**Comment by rmannibucau** [comment]
Updated according the comments.

**Comment by rmannibucau** [comment]
up?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 5]
```diff
@@ -209,7 +209,7 @@ public DoFnOperator(
 
     FlinkPipelineOptions flinkOptions = options.as(FlinkPipelineOptions.class);
 
-    this.maxBundleSize = flinkOptions.getMaxBundleSize();
+    this.maxBundleSize = options.getMaxBundleSize();
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: Seems like an unrelated change?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 6]
```diff
@@ -636,4 +648,49 @@ public void visitPrimitiveTransform(Node node) {
...
+    public String apply(final Map.Entry<String, Collection<PTransform<?, ?>>> input) {
+      final Collection<PTransform<?, ?>> values = instances.get(input.getKey());
+      return "- name=" + input.getKey() + ":\n"
+              + Joiner.on("\n").join(transform(values, TransformToMessage.INSTANCE));
+    }
+  }
+
+  private static class KeysExtractor implements
+          Function<Map.Entry<String, Collection<PTransform<?, ?>>>, String> {
+    private static final KeysExtractor INSTANCE = new KeysExtractor();
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: These INSTANCE variables are probably not worth the extra lines of code: just use `new KeysExtractor()` and `new IsUnique()`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java [hunk 7]
```diff
@@ -270,6 +270,11 @@
   String getJobName();
   void setJobName(String jobName);
 
+  @Description("The maximum number of elements in a bundle.")
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java]: Unrelated change, please revert

**Comment by rmannibucau** [comment]
Adressed comments, hopefully this time is the good one ;)

**Comment by jkff** [comment]
Thanks!
